### ISSUE TITLE
catkin build and run tests doesn't work properly since VS Code 1.69.1

### ISSUE THREAD

**Issue by FloMaile** [betwo/vscode-catkin-tools#46]
Referring to https://github.com/betwo/vscode-catkin-tools/issues/45.

CLI output doesn't show clearly the GTest output anymore with nice coloring indicating failed and passed test cases.

Additionally the Test Explorer doesn't work properly, as it runs into linker issues (not occurred with exactly the same setup, just immediately after the update of VS code)

**Comment by betwo** [comment]
Hi @FloMaile, thank you for reporting this.

Are you sure this is due to version 1.69.1? I.e. could you please check if downgrading to an older release of the extension also has the same issue?

In the mean time I'll try to reproduce it as soon as possible.

**Comment by betwo** [comment]
With the latest version, we have switched to vs code's native testing API (#42).
My assumption right now is that this caused the new linker and coloring issues with testing.

If downgrading the extension to 3.* does not show this behavior, that would confirm this theory.

I'm also currently working on a rework of the whole test handling to support other types of tests like ROS integration tests (#35). With that we can also tackle your linker errors.

**Comment by FloMaile** [comment]
I run into a linker issue with the version **3.3.6**.

```bash
error while loading shared libraries: liborocos-kdl.so.1.4: cannot open shared object file: No such file or directory
```

Coloring issues are still mixed present (some colorings work, some lines result in something like https://github.com/betwo/vscode-catkin-tools/issues/45)

**Comment by betwo** [comment]
Can we try to reproduce this in my setup?
Is there a minimal example you can share with me, or an open-source repository I can check out?

My first guess would be that this is a bug in `source`ing the workspace for the test execution, but so far I could not reproduce this.
Which shell are you using? And are there [errors in the debug output](https://github.com/betwo/vscode-catkin-tools/wiki/DebuggingInstalled)?

**Comment by phkehl** [comment]
I think I have the same problem (failed loading shared library). My setup is a docker container and vscode connected to that container. catkin_tools v4.3.1-prerelease. And using a .code-workspace file (File -> open workspace, not File -> open folder). Some observations:

- It detects the tests fine
- It appears that it compiles the tests correctly.
- In debug log there are only two messages:
    ```
    [Extension Host] 2022-08-11 06:38:03.168  DEBUG [extensions/betwo.b2-catkin-tools-4.3.1/dist/extension.js:2 t.Workspace.<anonymous>] updating with file /home/flip/ros/build/catkin_tools_prebuild/compile_commands.json
    [Extension Host] 2022-08-11 06:38:03.169  DEBUG [extensions/betwo.b2-catkin-tools-4.3.1/dist/extension.js:2  t.Workspace.startWatchingCompileCommandsFile] watching file /home/flip/ros/build/catkin_tools_prebuild/compile_commands.json 
    ```
- In output/catkin_tools console these:
    ```
    Reload ros
    Done loading folder ros / /home/flip/ros
    Initializing test adapter for workspace /home/flip/ros
    Loading tests
    ```
- And when I click the play (run test) button on one test it says:
    ```
    Running test(s): [object Object]
    ```
- After that the red circle icon appears in the test explorer.
- Shell used: bash (`"catkin_tools.shell": "bash"`)
- Clicking the debug test button: it starts spinning, says "running tests" in the bottom status bar, but nothing else happens for a while and no output in debug console

I've now tried opening the ROS workspace with just opening my folder (File -> open folder, instead of workspace):

- clicking the debug test button now leads to a terminal "cppdbg: some_test" being opened that complains:
    ```
    /home/flip/ros/devel/.private/some_ros_pkg/lib/some_ros_pkg/some_test: error while loading shared libraries: libsom_ros_pkg.so: cannot open shared object file: No such file or directory
    ```
- This is interesting as it also happens with manually crafted launch.json debug configs. However, there I can then just "source devel/setup.bash" in the terminal and start debugging again and then it works (as it reuses the same debug terminal again).
- I've been looking for ways to inject the env from setup.bash "globally" into vscode, but failed to find a way. As I'm connecting to a docker container, I cannot start vscode from inside the docker and a shell with setup.bash sourced, but instead have to connect to it. I guess this is what the catkin_tools.shell is all about?

Many thanks for your effort put into this!

**Comment by betwo** [comment]
Thank you for the detailed description and sorry again for the delay.

I failed to reproduce the issue after a lot of trials. Since I wanted to change the environment handling for the tests anyway, we might be lucky and this issue could be fixed "accidentally". 

Could you please test pre-release 4.3.2 with your setup?
This includes a few fixes that might have been the cause here:
* In some cases, the `/opt/ros/.../` workspace was not extended, causing invalid environments.
* There were also some issues with getting the environment variables of a given shell.

Do you have environment setup code in your `~/.${SHELL}rc` files?
With the latest version, any setup code that exports env variables should now be used.

**Comment by phkehl** [comment]
Hi. Thanks again. Unfortunately, I still cannot get it to work. Sourcing or not sourcing the setup.bash in .bashrc doesn't change the behaviour. I'd rather not source it in bashrc, as I'd like to be able to use multiple workspaces in different terminals. I.e., I manually source the right one for each terminal.

I've now tried again (with sourcing in bashrc) and tried to see what processes are running.

I think compiling the tests works fine but execution fails.

For example I can see this (what I think is the build command):
`bash --norc -c source /home/flip/ros/install/setup.bash > /dev/null 2 >& 1;pushd . > /dev/null; cd "/home/flip/ros"; cd "/home/flip/ros/build/some_package" .......`

I've not been able to see the command that executes it though.

When I try debugging, I still get the `/home/flip/ros/devel/.private/some_package/lib/some_package/some_test: error while loading shared libraries: libsome_package.so: cannot open shared object file: No such file or directory`

It's of course very well possible that some of my .bashrc or other configuration (Docker perhaps) messes with vscode or catkin_tools...

What do you mean by  "...any setup code that exports env variables should now be used."?

**Comment by betwo** [comment]
> Hi. Thanks again. Unfortunately, I still cannot get it to work. Sourcing or not sourcing the setup.bash in .bashrc doesn't change the behaviour. I'd rather not source it in bashrc, as I'd like to be able to use multiple workspaces in different terminals. I.e., I manually source the right one for each terminal.

That makes sense, I also do not source any workspace in my shell setup scripts and that should not be required in general.

> What do you mean by "...any setup code that exports env variables should now be used."?

Before we start executing the test, we run `env` in the user's shell after `source`ing the workspace's `setup.<shell>`. The output is then parsed and used as the environment variables for the process. (otherwise we would use the environment in which `code` was started.)
The same is also done for debugging tests.

I assume that this is also the place where the bug must be, somehow the environment is not correctly set up before running the test binary..

---

I just added a new test case with a dynamically built and linked library, that also worked in the docker-based integration test, very strange... Let's debug this more pragmatically now.

With `4.3.3`, there is now a new setting `catkin_tools.additionalEnvironmentVariables` for adding additional environment variables.
Could you please inject `LD_DEBUG=libs` and try to debug the test again?

```json
    "catkin_tools.additionalEnvironmentVariables": [
        "LD_DEBUG=libs"
    ]
```

Does this show any details about where `LD` is looking for before failing to find the `.so` file?

Also, does running the test (instead of debugging it) behave identically?

Does running the task `catkin_tools: run current package tests` also not work?

**Comment by phkehl** [comment]
Ah, some progress.. 

LD_DEBUG=libs helps. It appears the way catkin_tools runs the test is differently from manually running the test in the shell (I use the command `catkin test some_package`).

Via catkin_tools test explorer run test it looks for libraries in ~/ros/**install** but there is (almost) nothing in my ~/ros/install. I guess that is because my default ros workspace has `Install Space Layout:        None`?

After I run `catkin config --install` it changes to `Install Space Layout:        merged`. Whatever that means...

And now it appears that the tests do run (can I somewhere see their screen output?).
I can also re-run a test. It seems that this triggers a full recompilation. It takes quite a while to execute and CPU goes high with lots of gcc running. The build console output isn't visible anywhere, right?

Debugging sort of works. It starts gdb with the binary (and I no longer get the missing lib error in the terminal) but the debugger thingy in vscode says "unknown source". I can then hit the continue button and the test runs (and the test screen output shows in the terminal). But no stepping etc. The binary gdb uses is `~/ros/devel/.private/some_package/lib/some_package/some_test`.

I will have to play a bit more with this. But it seems it's more a setup problem in my workspace than a catkin_tools problem?

I should read up on catkin. I think I don't understand enough about this install space (install directory vs. devel directory, devel/.private, etc.) :-/

I'm not sure how I would run the `catkin_tools: run current package tests` task. I find no such thing anywhere. Should that be a build task? I can only see the build tasks I have manually configured in tasks.json (resp .code-workspace file).

It's late... I will experiment more with this.
Thanks a lot for your help!
Cheers!

**Comment by betwo** [comment]
Ah, yes, it looks like there is some issue of the install handling with the combination of what the extension does and what your workspace is configured for.

> Via catkin_tools test explorer run test it looks for libraries in ~/ros/install but there is (almost) nothing in my ~/ros/install. I guess that is because my default ros workspace has Install Space Layout: None?

That's kind of strange. The `install` space is only used if it exists.
So if you have not enabled `install: true` in the config (or via `catkin install`), then to folder should not exist at all.
Could you please check if the install space is created via the extension?
(I.e. clean your workspace (and make sure that the `install` folder is removed) and the rebuild and try again)

> After I run catkin config --install it changes to Install Space Layout: merged. Whatever that means...

AFAIK the install space is always `merged`, i.e. we cannot configure it to be `linked` as with the default `devel_layout`.
The difference is that with `merged`, all artifacts are copied into the devel/install directory, whereas with `linked`, artifacts stay in the `src`/`build` directory and are then only symlinked into `devel`.

> And now it appears that the tests do run (can I somewhere see their screen output?).
I can also re-run a test. It seems that this triggers a full recompilation. It takes quite a while to execute and CPU goes high with lots of gcc running. The build console output isn't visible anywhere, right?

You're right, it can currently take a while. I have not yet found a way to ensure that the tests are built before running them, without re-compiling first.
Do you have some ideas by chance?

You should be able to see the output in the test panel:
![Screenshot from 2022-09-03 10-58-07](https://user-images.githubusercontent.com/4789818/188263723-c9db8a39-95b6-4ce7-96c7-4b68b81d83fa.png)
![Screenshot from 2022-09-03 10-58-15](https://user-images.githubusercontent.com/4789818/188263728-73e51378-1616-466a-8c7b-478cc5a4ea8c.png)

> Debugging sort of works. It starts gdb with the binary (and I no longer get the missing lib error in the terminal) but the debugger thingy in vscode says "unknown source". I can then hit the continue button and the test runs (and the test screen output shows in the terminal). But no stepping etc. The binary gdb uses is ~/ros/devel/.private/some_package/lib/some_package/some_test.

Ah, I can reproduce this in a `Release` build, (for which this behavior makes sense, there is no debug information.)
You can enable debugger support with
```
-DCMAKE_BUILD_TYPE=RelWithDebInfo
```
or
```
-DCMAKE_BUILD_TYPE=Debug
```
in your catkin profile, depending on whether you want code optimization performed.

---

> I'm not sure how I would run the catkin_tools: run current package tests task. I find no such thing anywhere. Should that be a build task? I can only see the build tasks I have manually configured in tasks.json (resp .code-workspace file).

Hmm, yes, that should be a build task provided by the catkin_tools extension
![catkin_tasks](https://user-images.githubusercontent.com/4789818/188263978-4c617cd7-098d-4b8f-ac80-161b846b93a7.png)
![catkin_task_build_and_run_test](https://user-images.githubusercontent.com/4789818/188263980-5836690e-7d59-4440-a9dd-3af0839a3b39.png)

---

Do you think adding GUI support for enabling the cmake debug modes via the extension would help here?

**Comment by phkehl** [comment]
Hi, many thanks again. I think I have now something working here. Maybe my ROS workspace was all foobar. I've started a new one from scratch (removed all folders incl. .catkin_config) and things seem to work now.

I did `catkin init` and `catkin build my_package`. After that no install folder exists and `catkin config` says "Install Space Layout: None".

Then I run the tests via text explorer, which all compile and run fine. But now an install folder has appeared with my_package stuff in it. Catkin config still says "None", though.

Unfortunately, that catkin and cmake stuff is in many parts still a big mystery to me. I don't think I can be of much help in finding out what's going on here or in finding a way to compile-only-if-necessary. I'm more than happy to beta test any idea you may have!

I think one thing that confused me a lot when running the tests was the need to wait for the compilation. I may have been too impatient waiting for the results! And somehow I have never noticed that "Show Output" button. Thanks for pointing me to it!

So I think this solves running tests for me.

Another good hint from you is regarding CMAKE_BUILD_TYPE. Some of the CMakeLists.txt in the packages I use hardcode CMAKE_BUILD_TYPE=Release. Even though I typically set the catkin workspace for CMAKE_BUILD_TYPE=Debug, those packages with hardcoded CMAKE_BUILD_TYPE will not build with Debug. I should probably change those to RelWithDebInfo. Or don't hardcode it at all (need to talk to my colleagues regarding that.. they're having some problems elsewhere with Debug builds for some reason and hence they have hardcoded Release builds.. :-/)

I think this too was more an issue with my setup than with catkin_tools.

Regarding the build tasks (CTRL-SHIFT-B, right?). I still only see those tasks defined in my .code-workspace. But that might be a vscode issue. Those .code-workspace things are not working 100% correct in other places, too. This is not an issue for me. I'm a switch-to-console-and-run-build-commands-manually person :-)

More GUI help from catkin_tools is much appreciated. Similar to the popup regarding CMAKE_EXPORT_COMPILE_COMMANDS it could perhaps hint at CMAKE_BUILD_TYPE not being Debug or RelWithDebInfo?

Also the catkin_tools.additionalEnvironmentVariables config was very useful to debug things.

I believe we can now say that catkin_tools works as intended to run and debug tests.

Let me know if you'd like me to tests other things

Many thanks for your help!

**Comment by betwo** [comment]
Sorry, I totally lost track of this and had little time to continue the work.

Did you find a way to solve your open problems? And did anything relate to the extension rather than catkin/cmake setup?

I'd close the issue now and add the GUI help topic to the "backlog". Thank you for all the beta testing!
Please feel free to open up new issues :)

**Comment by phkehl** [comment]
Yes, it's working now.
I'm still a bit puzzled by the complexity and occasional slowness of catkin and the builds, but that would be another issue in another repo... :-)
Many thanks again for your help!
